### PR TITLE
state: add an 'IsTimely' method to implement the 'timely' check for proposer-based timestamps

### PR DESCRIPTION
**Pull request by williambanfield**
This pull requests implements the 'timely' as outlined in the proposer-based timestamps specification.

For a proposed block timestamp to be considered timely, it must satisfy the following inequality:

```
proposedBlockTime < validatorLocalTime + Precision + MsgDelay && proposedBlockTime > validatorLocaltime - Precision.
```
Where `Precision` and `MsgDelay` are system parameters where `Precision` bounds the acceptable clock drift between any two correct process and `MsgDelay` bounds the acceptable delay for full transmitting a message between any two processes.

part of #6942

### CODE REVIEW

**Review by tychoish — APPROVED** [review]

File: types/block.go [hunk 1]
```diff
@@ -94,6 +95,25 @@ func (b *Block) ValidateBasic() error {
...
+// IsTimely validates that the block timestamp is 'timely' according to the proposer-based timestamp algorithm.
+// To evaluate if a block is timely, its timestamp is compared to the local time of the validator along with the configured
+// Precision and MsgDelay parameters.
+// Specifically, a proposed block timestamp is considered timely if it is satisfies the following inequalities:
+//
+// proposedBlockTime > validatorLocaltime - Precision && proposedBlockTime < validatorLocalTime + Precision + MsgDelay.
+//
+// For more information on the meaning of 'timely', see the proposer-based timestamp specification:
+// https://github.com/tendermint/spec/tree/master/spec/consensus/proposer-based-timestamp
+func (b *Block) IsTimely(clock tmtime.Source, p TimestampParams) bool {
```
Comment (cmwaters) [inline types/block.go:107]: Don't we measure timeliness when we receive the proposal or when we receive the block. I thought it was based on the proposal

File: types/params.go [hunk 2]
```diff
@@ -86,6 +93,7 @@ func DefaultConsensusParams() *ConsensusParams {
 }
 
 // DefaultBlockParams returns a default BlockParams.
+// TODO (@wbanfield): add link to proposer-based timestamp spec when completed.
```
Comment (cmwaters) [inline types/params.go:96]: Why did you add this link here?

File: types/params.go [hunk 3]
```diff
@@ -75,6 +75,13 @@ type VersionParams struct {
 	AppVersion uint64 `json:"app_version"`
 }
 
+// TimestampParams influence the validity of block timestamps.
+type TimestampParams struct {
```
Comment (cmwaters) [inline types/params.go:79]: Do we want to add a default?

Comment (williambanfield) [reply types/params.go:81]: Set some basics and will follow up in https://github.com/tendermint/tendermint/issues/7202

File: types/params.go [hunk 4]
```diff
@@ -116,6 +125,16 @@ func DefaultVersionParams() VersionParams {
 	}
 }
 
+func DefaultTimestampParams() TimestampParams {
+	// TODO(@wbanfield): Determine experimental values for these defaults
+	// https://github.com/tendermint/tendermint/issues/7202
+	return TimestampParams{
+		Precision: 2 * time.Second,
+		Accuracy:  500 * time.Millisecond,
```
Comment (cason) [inline types/params.go:133]: Observe that if Accuracy = 0.5ms, two clocks might differ by at most Precision = 1ms.

File: types/proposal.go [hunk 5]
```diff
@@ -79,6 +79,25 @@ func (p *Proposal) ValidateBasic() error {
 	return nil
 }
 
+// IsTimely validates that the block timestamp is 'timely' according to the proposer-based timestamp algorithm.
+// To evaluate if a block is timely, its timestamp is compared to the local time of the validator along with the
+// configured Precision and MsgDelay parameters.
+// Specifically, a proposed block timestamp is considered timely if it is satisfies the following inequalities:
+//
+// proposedBlockTime > validatorLocaltime - Precision && proposedBlockTime < validatorLocalTime + Precision + MsgDelay.
```
Comment (cason) [inline types/proposal.go:87]: > proposedBlockTime > validatorLocaltime - Precision

Are you sure that this is correct?

File: types/proposal.go [hunk 6]
```diff
@@ -79,6 +79,25 @@ func (p *Proposal) ValidateBasic() error {
...
+//
+// For more information on the meaning of 'timely', see the proposer-based timestamp specification:
+// https://github.com/tendermint/spec/tree/master/spec/consensus/proposer-based-timestamp
+func (p *Proposal) IsTimely(clock tmtime.Source, tp TimestampParams) bool {
+	lt := clock.Now()
+	lhs := lt.Add(-tp.Precision)
+	rhs := lt.Add(tp.Precision).Add(tp.MsgDelay)
+	if lhs.Before(p.Timestamp) && p.Timestamp.Before(rhs) {
```
Comment (cason) [inline types/proposal.go:95]: Maybe just a detail, but I would prefer having something like "a.After(b) and a.Before(c)" for more clarity.

File: types/proposal_test.go [hunk 7]
```diff
@@ -191,3 +192,65 @@ func TestProposalProtoBuf(t *testing.T) {
...
+		proposalTime time.Time
+		localTime    time.Time
+		precision    time.Duration
+		msgDelay     time.Duration
+		expectTimely bool
+	}{
+		{
+			// Checking that the following inequality evaluates to true:
+			// 1 - 2 < 0 < 1 + 2 + 1
+			name:         "basic timely",
```
Comment (cason) [inline types/proposal_test.go:210]: For me, the "basic" case, meaning the most common, is:
 * The proposal timestamp is "for sure" in the past, so: proposalTime + Precision < localTime - Precision
 * The receiving time of the proposal is "fore sure" within the acceptable window, so: localTime + Precision < proposalTime + msgDelay - Precision

File: types/proposal_test.go [hunk 8]
```diff
@@ -191,3 +192,65 @@ func TestProposalProtoBuf(t *testing.T) {
...
+			proposalTime: genesisTime,
+			localTime:    genesisTime.Add(1 * time.Nanosecond),
+			precision:    time.Nanosecond * 2,
+			msgDelay:     time.Nanosecond,
+			expectTimely: true,
+		},
+		{
+			// Checking that the following inequality evaluates to false:
+			// 3 - 2 < 0 < 3 + 2 + 1
+			name:         "local time too large",
```
Comment (cason) [inline types/proposal_test.go:220]: Well, is not the local time that is large, but the Proposal took too long to be received.

File: types/proposal_test.go [hunk 9]
```diff
@@ -191,3 +192,65 @@ func TestProposalProtoBuf(t *testing.T) {
...
+			proposalTime: genesisTime,
+			localTime:    genesisTime.Add(3 * time.Nanosecond),
+			precision:    time.Nanosecond * 2,
+			msgDelay:     time.Nanosecond,
+			expectTimely: false,
+		},
+		{
+			// Checking that the following inequality evaluates to false:
+			// 0 - 2 < 2 < 2 + 1
+			name:         "proposal time too large",
```
Comment (cason) [inline types/proposal_test.go:230]: Maybe a more clear name would be: "proposal timestamp in the future"

File: types/proposal_test.go [hunk 10]
```diff
@@ -191,3 +192,65 @@ func TestProposalProtoBuf(t *testing.T) {
...
+		expectTimely bool
+	}{
+		{
+			// Checking that the following inequality evaluates to true:
+			// 1 - 2 < 0 < 1 + 2 + 1
+			name:         "basic timely",
+			proposalTime: genesisTime,
+			localTime:    genesisTime.Add(1 * time.Nanosecond),
+			precision:    time.Nanosecond * 2,
+			msgDelay:     time.Nanosecond,
```
Comment (cason) [inline types/proposal_test.go:214]: I would expect msgDelay >> Precision.

**Review by cason — COMMENTED** [review]
For me is not clear whether the corner cases are defined or tested properly. I also think that the main `isTimely` method could be easier to read, by using Before() and After() and Add() and Sub() `time` methods.

File: types/block.go (hunk 1 above)
Comment (cason) [reply types/block.go]: In terms of the implementation, this also seems easier. If a Proposal is not timely, when it is received and processed at first by the consensus implementation, there is no point in "asking for" and receiving the corresponding BlockParts. In addition, if the timely call is made for the block, when in the implementation will it be called? This is relevant, as different calls to timely may return true and the false, depending on when they are invoked.

File: types/params.go (hunk 3 above)
Comment (cason) [reply types/params.go:81]: These would be Synchronous or Timing parameters, wouldn't them?

File: types/proposal.go (hunk 6 above)
Comment (williambanfield) [reply types/proposal.go:95]: `.After` in Go is actually just a little syntactic sugar around `Sleep`. It's definitely annoying that it's named so similarly to `.Before`, but the functions don't do the same thing.

**Comment by williambanfield** [comment]
> For me is not clear whether the corner cases are defined or tested properly. I also think that the main `isTimely` method could be easier to read, by using Before() and After() and Add() and Sub() `time` methods.

`.Sub` produces a `time.Duration`, whereas what we need is a `time.Time`. `Add` produces a `time.Time`, another sort of annoying difference in similarly named functions.

File: types/proposal.go (hunk 5 above)
Comment (williambanfield) [reply types/proposal.go:87]: I believe so, according to this description in the spec https://github.com/tendermint/spec/blob/master/spec/consensus/proposer-based-timestamp/pbts-algorithm_001_draft.md#pbts-reception-step0
